src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -1759,18 +1759,123 @@ void grpc_chttp2_ack_ping(grpc_chttp2_transport* t, uint64_t id) {
   }
 }
 
+namespace {
+
+// Fire and forget (deletes itself on completion). Does a graceful shutdown by
```

Comment:
My understanding of this type is that it keeps the transport alive an extra 20 seconds unconditionally - I think this is problematic -- if we signal GOAWAY and wait one RTT, then once the number of requests on a transport drops to zero we are able to complete this dance deterministically and release the (significant) memory used by the transport much earlier.

Reply:
No, it's not an unconditional 20 second wait. It only waits 20 seconds if the peer does not respond to the ping. If there are no streams and the RTT completes with a ping ack, then we would immediately send the final goaway and close the transport. (Some of our tests fail on an unconditional wait.)

Reply:
I missed the grpc_timer_cancel last night! That makes sense... let me take another pass on this.